sdk/compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs

```diff
@@ -288,20 +292,22 @@ getExternalPackages file = do
         map LF.dalfPackagePkg (Map.elems pkgMap) <> map LF.dalfPackagePkg (Map.elems stablePackages)
 
 -- Generates and type checks the DALF for a module.
-generateDalfRule :: Rules ()
-generateDalfRule =
+generateDalfRule :: Options -> Rules ()
+generateDalfRule opts =
     define $ \GenerateDalf file -> do
         lfVersion <- getDamlLfVersion
         WhnfPackage pkg <- use_ GeneratePackageDeps file
         pkgs <- getExternalPackages file
         let world = LF.initWorldSelf pkgs pkg
         rawDalf <- use_ GenerateRawDalf file
+        upgradedPackage <- join <$> useNoFile ExtractUpgradedPackage
```

Comment:
This doesn't read UpgradeInfo from the opts, it reads the DAR to which the upgradeDar field in opts refers. I do it as a rule so that it can be cached and we only read in the DAR once - if we specified it as a function then the DAR would be read as many times as there are modules.

I'd rather not clutter the rule by returning extra UpgradeInfo from it, because UpgradeInfo is needed even when there is no upgradeDar, and because opts is the default way of passing around build config in Daml.Rules.